File: frequencyTopic2020/hash/subarraySum.cc
#include <iostream>
#include <vector>

using namespace std;

class Solution {
 public:
  int subarraySum(vector<int>& nums, int k) {
    if (nums.empty()) {
      return 0;
    }
    return subByIdx(nums, 0, k);
  }

  int subByIdx(vector<int>& nums, int i, int k) {
    int res = 0;
    while (i < nums.size()) {
      int j = i + 1;
      int sum = nums[i];
      while (j < nums.size() && sum + nums[j] != k) {
        j++;
        sum += nums[j];
      }
      if (sum == k) {
        if (j < nums.size() - 1) {
          res += 1 + subByIdx(nums, j, 0);
        }
        res++;
      }
      i++;
    }
    return res;
  }
};

File: rush/hash/subarraySum.go
package main

/*
func subarraySum(nums []int, k int) int {
    
    if len(nums) == 0 {
        return 0
    }

    var res int
    for i := 0; i < len(nums); i++ {
        sum := nums[i]
        for j := i + 1; j < len(nums); j++ {
            sum += nums[j]
            if sum == k {
                res++
            }
        }
    }

    return res
}
*/

func subarraySum(nums []int, k int) int {

    if len(nums) == 0 {
        return 0
    }

    record := map[int]int{0: 1}
    var pre, res int
    for i := range nums {
        pre += nums[i]
        if f, found := record[pre - k]; found {
            res += f
        }
        record[pre]++ // why !found record[pre] = 1 !ac?
    }

    return res
}
